examples/step-57/step-57.cc

```diff
@@ -0,0 +1,879 @@
+        std::cout << "viscosity= " << viscosity << std::endl;
+        std::cout << "*****************************************" << std::endl;
+
+        while ((first_step || (current_res > tolerance)) && outer_iteration < max_iteration)
+          {
+            if (first_step)
+              {
+                setup_dof();
+                initialize_system();
+                evaluation_point = present_solution;
+                assemble_matrix(first_step);
+                assemble_rhs(first_step);
+                solve(first_step);
+                present_solution = newton_update;
+                nonzero_constraints.distribute(present_solution);
+                first_step = false;
+                evaluation_point = present_solution;
+                assemble_rhs(first_step);
+                current_res = system_rhs.l2_norm();
+                std::cout << "******************************" << std::endl;
+                std::cout << " The residual of initial guess is " << current_res << std::endl;
+                std::cout << " Initialization complete!  " << std::endl;
+                last_res = current_res;
+              }
+
+            else
+              {
+                evaluation_point = present_solution;
+                assemble_matrix(first_step);
+                if (outer_iteration == 0)
+                  assemble_rhs(first_step);
+                solve(first_step);
+
+                for (double alpha = 1.0; alpha > 1e-5; alpha *= 0.5)
+                  {
+                    evaluation_point = present_solution;
+                    evaluation_point.add(alpha, newton_update);
+                    assemble_rhs(first_step);
+                    current_res = system_rhs.l2_norm();
+                    std::cout << " alpha = " << std::setw(6) << alpha << std::setw(0)
+                              << " res = " << current_res << std::endl;
+                    if (current_res < last_res)
+                      break;
+                  }
+
+                {
+                  present_solution = evaluation_point;
+                  nonzero_constraints.distribute(present_solution);
+                  std::cout << " ---- Iteration " << outer_iteration << " residual: " << current_res << std::endl;
+                  last_res = current_res;
+                }
+
+              }
+            ++outer_iteration;
+
+            if (output_result)
+              {
+                output_results (max_iteration*refinement+outer_iteration);
+
+                if (current_res <= tolerance)
+                  process_solution(refinement);
+              }
+          }
+
+        if (refinement < n_refinements)
+          {
+            refine_mesh();
+            std::cout << "*****************************************" << std::endl
+                      << "        Do refinement ------   " << std::endl;
+          }
+      }
+
+
+  }
+
+  // @sect4{StationaryNavierStokes::refine_mesh}
```

Comment:
We have typically put higher level functions below lower level functions. Since the previous function calls this one, maybe move this one above `newton_iteration()`?